---
title: 'RNAmodR: AlkAnilineSeq'
author: "Felix G.M. Ernst"
date: "`r Sys.Date()`"
abstract: >
  detection of m7G, m3C and D modification by AlkAnilineSeq
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{RNAmodR.AlkAnilineSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown(css.files = c('custom.css'))
```

# Introduction









```{r, echo = FALSE}
suppressPackageStartupMessages({
  library(rtracklayer)
  library(Rsamtools)
  library(GenomicFeatures)
  library(RNAmodR.AlkAnilineSeq)
})
```
```{r, eval = FALSE}
library(rtracklayer)
library(Rsamtools)
library(GenomicFeatures)
library(RNAmodR.AlkAnilineSeq)
```

# Example workflow

Loading files
```{r}
RNAmodR.files <- RNAmodR.Data::RNAmodR.files
annotation <- GFF3File(RNAmodR.files[["example_AAS.gff3"]])
sequences <- FaFile(RNAmodR.files[["example_AAS.fasta"]])
files <- list("SampleSet1" = c(treated = RNAmodR.files[["example_wt_1.bam"]],
                               treated = RNAmodR.files[["example_wt_2.bam"]],
                               treated = RNAmodR.files[["example_wt_3.bam"]]),
              "SampleSet2" = c(treated = RNAmodR.files[["example_bud23_1.bam"]],
                               treated = RNAmodR.files[["example_bud23_2.bam"]]),
              "SampleSet3" = c(treated = RNAmodR.files[["example_trm8_1.bam"]],
                               treated = RNAmodR.files[["example_trm8_2.bam"]]))
```

Analysis of data

```{r}
msaas <- ModSetAlkAnilineSeq(files, annotation = annotation, sequences = sequences)
```

# Visualizing the results

```{r}

```

Heatmap

```{r}

```

sequence specific result

```{r}
visualizeData(msaas, "1", from = 1550L, to = 1600L)
```

# Session info

```{r, sessioninfo}
sessioninfo::session_info()
```

# References
